numpy/_core/src/multiarray/refcount.c

```diff
@@ -56,6 +56,61 @@ PyArray_ClearBuffer(
 }
 
 
+/*
+ * Helper function to zero an array buffer.
+ *
+ * Here "zeroing" means an abstract zeroing operation, which for an
+ * array of references might be something more complicated than
+ * zero-filling the buffer.
+ *
+ * Failure (returns -1) indicates some sort of programming or logical
+ * error and should not happen for a data type that has been set up
+ * correctly. In principle a sufficiently weird dtype might run out of
+ * memory but in practice this likely won't happen.
+ */
+NPY_NO_EXPORT int
+PyArray_ZeroContiguousBuffer(
+        PyArray_Descr *descr, char *data,
+        npy_intp stride, npy_intp size, int aligned)
+{
+    NPY_traverse_info zero_info;
+    NPY_traverse_info_init(&zero_info);
+    /* Flags unused: float errors do not matter and we do not release GIL */
+    NPY_ARRAYMETHOD_FLAGS flags_unused;
+    PyArrayMethod_GetTraverseLoop *get_fill_zero_loop =
+            NPY_DT_SLOTS(NPY_DTYPE(descr))->get_fill_zero_loop;
+    if (get_fill_zero_loop != NULL) {
+        if (get_fill_zero_loop(
+                    NULL, descr, aligned, descr->elsize, &(zero_info.func),
+                    &(zero_info.auxdata), &flags_unused) < 0) {
+            goto fail;
+        }
+    }
+    else {
+        npy_intp nbytes;
+        if (!npy_mul_sizes_with_overflow(&nbytes, size, stride)) {
```

Comment:
I would be happier with _not_ checking it here, TBH (not because I mind it here, but because I don't want to clutter more code with cargo-culted checks).
A valid NumPy buffer cannot overflow this so if this happens there would be a missing check earlier.  Maybe an `assert` instead if you like to clarify it?

After all, just like ufuncs/other functions the `zero_info.func` is not and should not be required to check this as well.